pkg/storage/store_test.go

```diff
@@ -858,3 +860,94 @@ func buildTestStreams(labels string, tr timeRange) logproto.Stream {
+		From:      chunk.DayTime{Time: model.Now().Add(-12 * time.Hour)},
+		IndexType: "second",
+	})
+	assert.Equal(t, 1, ActivePeriodConfig(cfg))
+
+	// add a newer PeriodConfig in the future which should not be considered
+	cfg.Configs = append(cfg.Configs, chunk.PeriodConfig{
+		From:      chunk.DayTime{Time: model.Now().Add(time.Hour)},
+		IndexType: "third",
+	})
+	assert.Equal(t, 1, ActivePeriodConfig(cfg))
+}
+
+func TestUsingBoltdbShipper(t *testing.T) {
+	var cfg SchemaConfig
+
+	// just one PeriodConfig in the past using boltdb-shipper
+	cfg.Configs = []chunk.PeriodConfig{{
+		From:      chunk.DayTime{Time: model.Now().Add(-24 * time.Hour)},
+		IndexType: "boltdb-shipper",
+	}}
+	assert.Equal(t, true, UsingBoltdbShipper(cfg))
+
+	// just one PeriodConfig in the past not using boltdb-shipper
+	cfg.Configs[0].IndexType = "boltdb"
+	assert.Equal(t, false, UsingBoltdbShipper(cfg))
+
+	// add a newer PeriodConfig in the future using boltdb-shipper
+	cfg.Configs = append(cfg.Configs, chunk.PeriodConfig{
+		From:      chunk.DayTime{Time: model.Now().Add(time.Hour)},
+		IndexType: "boltdb-shipper",
+	})
+	assert.Equal(t, true, UsingBoltdbShipper(cfg))
+}
+
+func TestSchemaConfig_Validate(t *testing.T) {
+	var cfg SchemaConfig
+
+	// NOT using boltdb-shipper. SHOULD NOT ERROR.
+	cfg.Configs = []chunk.PeriodConfig{{
+		From:      chunk.DayTime{Time: model.Now().Add(-24 * time.Hour)},
+		IndexType: "boltdb",
+		Schema:    "v9",
+		IndexTables: chunk.PeriodicTableConfig{
+			Period: 7 * 24 * time.Hour,
+		},
+	}}
+	require.NoError(t, cfg.Validate())
+
+	// current config boltdb-shipper with 7 days periodic config, without future index type changes. SHOULD ERROR.
+	cfg.Configs[0].IndexType = "boltdb-shipper"
+	require.EqualError(t, cfg.Validate(), currentBoltdbShipperNon24HoursErr.Error())
+
+	// current config boltdb-shipper with 1 day periodic config, without future index type changes. SHOULD NOT ERROR.
+	cfg.Configs[0].IndexTables.Period = 24 * time.Hour
+	require.NoError(t, cfg.Validate())
+
+	// current config boltdb-shipper with 7 days periodic config, upcoming config not boltdb-shipper. SHOULD NOT ERROR.
+	cfg.Configs[0].IndexTables.Period = 7 * 24 * time.Hour
+	cfg.Configs = append(cfg.Configs, chunk.PeriodConfig{
+		From:      chunk.DayTime{Time: model.Now().Add(time.Hour)},
+		IndexType: "boltdb",
+		Schema:    "v9",
+		IndexTables: chunk.PeriodicTableConfig{
+			Period: 7 * 24 * time.Hour,
+		},
+	})
+	require.NoError(t, cfg.Validate())
+
+	// current and upcoming config boltdb-shipper with 7 days periodic config. SHOULD ERROR.
+	cfg.Configs[1].IndexType = "boltdb-shipper"
+	require.EqualError(t, cfg.Validate(), upcomingBoltdbShipperNon24HoursErr.Error())
+
+	// current config not boltdb-shipper, upcoming config boltdb-shipper with 7 days periodic config. SHOULD ERROR.
+	cfg.Configs[0].IndexType = "boltdb"
+	require.EqualError(t, cfg.Validate(), upcomingBoltdbShipperNon24HoursErr.Error())
```

Comment:
This kind of test is a great git for go table tests, your coverage here is good but I think table tests make this a lot easier to read, and would additionally make the tests a little more obvious.

I had to look for a bit to understand in the last 2 cases exactly which config was getting manipulated.

Table tests here would also define new config structs for each tests which is a easier to read and less error prone (hopefully) because we wouldn't be re-using the same struct.

I think we should look to use table tests as much as possible, this has been kind of difficult sell for me initially but after reading through a bunch of tests @cyriltovena has created for PR's I've become a real believer.

Reply:
looks great, thank you!